### PR TITLE
System auth cache

### PR DESCRIPTION
**Pull request by zahnno**
Added to improve performance on the system auth service decorator.

### CODE REVIEW

File: src/foam/nanos/auth/SystemAuthService.js [hunk 1]
```diff
@@ -11,17 +11,35 @@ foam.CLASS({
...
         Group group = (Group) x.get("group");
-        boolean isSystem = user != null && user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID;
-        return isSystem || group != null && group.implies(x, new AuthPermission("*")) || getDelegate().check(x, permission);
+        if ( user != null && user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID ) {
+          return true;
+        }
+        // Response from group implies is cached to maintain intended performance.
+        if ( group != null ) {
+          Session session = x.get(Session.class);
+          Boolean isSuper = (Boolean) session.getContext().get(CACHE_KEY);
```
Comment (kgrgreer) [inline src/foam/nanos/auth/SystemAuthService.js]: boolean instead of Boolean

File: src/foam/nanos/auth/SystemAuthService.js [hunk 2]
```diff
@@ -11,17 +11,35 @@ foam.CLASS({
...
+          if ( isSuper == null ) {
+            isSuper = group.implies(x, new AuthPermission("*"));
+            session.setContext(session.getContext().put(CACHE_KEY, isSuper));
+          }
+          return isSuper || getDelegate().check(x, permission);
```
Comment (kgrgreer) [inline src/foam/nanos/auth/SystemAuthService.js]: This seems fragile if the user changes their group. Could we just keep an instance variable with a ConcurrentHashMap to cache super groups?

File: src/foam/nanos/auth/SystemAuthService.js (hunk 1 above)
Comment (zahnno) [reply src/foam/nanos/auth/SystemAuthService.js]: Checking for null below - we want Booleans object/reference behaviour unless you have another idea

File: src/foam/nanos/auth/SystemAuthService.js (hunk 2 above)
Comment (zahnno) [reply src/foam/nanos/auth/SystemAuthService.js]: > This seems fragile if the user changes their group.

Can you elaborate your concern?

> Could we just keep an instance variable with a ConcurrentHashMap to cache super groups?

I don't know - When a request to the server is made does the system create a new instance of each java class exclusive to the user and is maintained during their session?

**Pull request merged**